luanti-org/luanti
# SDL: Use scancodes for keybindings

This PR makes Minetest use scancodes for keybindings.

- Goal of the PR
  Make Minetest use scancodes (instead of keycodes) for key bindings.
- How does the PR work?
  The SDL scancode is passed to `SKeyInput::SystemKeyCode`. The `KeyPress` class is modified to be basically a wrapper around this.
- Does it resolve any reported issue?
  - Fixes #14940
  - Fixes #13904 (if not, this would be an upstream issue)
  - Fixes #15747
  - Resolves #12556 (relevant code path removed)
  - Resolves #8781 (keycode-based syntax deprecated)
- Does this relate to a goal in [the roadmap](https://github.com/minetest/minetest/blob/master/doc/direction.md)?
  Yes - SDL.
- If not a bug fix, why is this PR needed? What usecases does it solve?
  This was discussed recently.

Related: #12507 (new default key), #14545, #14874 (allows people to continue using e.g. Shift+7 if they want to).

Documentation update: luanti-org/docs.luanti.org#211

## To do

This PR is ready for review.

- [x] Switch to scancodes for SDL. Address regressions:
  - [x] Fix: Escape and `AC_BACK` keys not working.
  - [x] Fix: (Android) Cannot close formspec when a text field is focused.
  - [x] Fix: actions without keybindings are bound to the left mouse button.
  - [x] Properly implement scancodes (using `std::variant`) for keypresses from non-keyboard input (mouse, joystick, etc.)
  - [x] Convert `KEY_OEM_*` keys to scancodes directly. (See note below)
  The current approach no longer requires changing non-SDL `IrrlichtDevices`.
- Partial rework of `KeyPress` and event handling
  - [x] Rework `KeyPress`
  - [x] Allow binding to keys with keychars (keycodes) not in the lookup table.
  - [x] Allow binding to keys with no keycode (e.g. dead keys).
  - [x] Change touchscreen GUI to send events with scancodes.
- Keybinding settings
  - [x] Add syntax for scancode settings. (see notes)
- Followup PRs:
  - Allow using secondary keybindings: #15577
  - Allow keybindings with modifer keys: #14874
  - **Important:** Warn users about new keybindings.
  - After removing non-SDL `IrrlichtDevice`s:
    - Expose SDL headers to Luanti.
    - Implement `std::hash<KeyPress>` and replace `KeyList` with `std::unordered_set<KeyPress>`
    - Implement SDL-specific scancode names (i.e. `SDL_SCANCODE_x`). (Useful for #12488) Note that this requires Luanti (not just Irrlicht) to have access to SDL headers.
    - Fix unittests

## How to test
* Set keybindings as usual or use existing keybinding settings. (In particular, test binding to keys such as backtick, `ü`, and dead keys). Observe that the keybinding (in terms of scancodes) is the same after switching to a different layout.
* Observe that both the Escape key and the Android back key work as the same key.

## Notes
* The scancode-based keybinding syntax introduced in this PR is only used (and enabled) for SDL builds.
* Unittests have been disabled:
  * The behavior of `KeyPress` is fully dependent on the Irrlicht device.
  * Scancode/keycode conversions for SDL require an active `IrrlichtDevice` which is not available in unittests.
* The "press Shift to bind to symbols" feature has been removed as it does not work with SDL scancodes. Note that this causes existing configurations to break *even when Minetest is built without SDL*.
* The Irrlicht source code suggests that the various OEM keys refer to phsical keys (similar to SDL scancodes), but on my non-SDL X11 build it appears that the _exact_ OEM key still depends on the keyboard layout.
* https://github.com/minetest/minetest/pull/14964#issuecomment-2299626800
* This PR uses scancodes for individual keys. In particular, left shift and right shift are treated as separate keys. *Addressing this requires a separate/followup PR that implements secondary keybindings.*
* We will end up (one way or another) requiring requiring some users to rebind their keys regardless of what we choose as the default. Scancode-based defaults basically break keyboard layouts where a key is located in a different position than its US counterpart, while keycode-based configurations would break on layouts where a key is not present in the **unshifted** position. (https://github.com/minetest/minetest/pull/14964#issuecomment-2559974639 and https://irc.minetest.net/minetest-dev/2024-12-26#i_6230893) Using a mixture of both can result in key conflicts (https://irc.minetest.net/minetest-dev/2024-11-10#i_6216414)
